corehq/form_processor/models/cases.py

```diff
@@ -838,13 +839,14 @@ def save(self, *args, with_tracked_models=False, **kw):
         except DatabaseError as e:
             raise CaseSaveError(e)
 
-    def delete(self, leave_tombstone=False):
-        if not leave_tombstone and not settings.UNIT_TESTING:
-            raise ValueError(
-                'Cannot delete case without leaving a tombstone except during testing, domain deletion or '
-                'when deleting system cases')
-        if leave_tombstone:
-            create_deleted_sql_doc(self.case_id, 'form_processor.CommCareCase', self.domain, self.deleted_on)
+    def delete(self, leave_tombstone=True):
+        if not settings.UNIT_TESTING:
+            if not leave_tombstone:
+                raise ValueError(
+                    'Cannot delete case without leaving a tombstone except during testing, domain deletion or '
+                    'when deleting system cases')
+            if leave_tombstone:
+                create_deleted_sql_doc(self.case_id, 'form_processor.CommCareCase', self.domain, self.deleted_on)
```

Comment:
Given the above condition, `if leave_tombstone` will always evaluate to true, right?
```suggestion
            create_deleted_sql_doc(self.case_id, 'form_processor.CommCareCase', self.domain, self.deleted_on)
```

Reply:
Ha this threw me for a loop. Is it the case that having `leave_tombstone` as a parameter was only necessary for Domains because the tombstone was still a Domain-Deleted couch doc that could "get in the way" of tests? Whereas now that the tombstone is just a separate SQL table, it can't negatively impact tests so we can just always create a tombstone when calling `case.delete()`?

Have we thought through what hard deleting cases will look like? The simple case is when a case has been deleted and a tombstone exists, so now we just have to delete the tombstone, but what if a domain deletion is requested on not yet deleted cases? Are we back to a two step process where we call `case.delete()`, a tombstone is created, and then we have to also delete that tombstone? That seems like where it would be nice be able to set `leave_tombstone` to False right? Which is impossible with the existing code?

Reply:
> Whereas now that the tombstone is just a separate SQL table, it can't negatively impact tests so we can just always create a tombstone when calling case.delete()?

Case.delete() can't negatively impact tests because there's an explicit check that excludes calls from tests from making tombstones. 

> Are we back to a two step process where we call case.delete(), a tombstone is created, and then we have to also delete that tombstone?

Domain deletion handles case deletion through bulk deletion, which doesn't touch `CommCareCase.delete()` and so currently can't create tombstones. I'm a little confused on how to move forward here because it seems that both you and Daniel agree that it would be best to not create tombstones for case/form deletion via domain deletion and system deletion (specifically user owner mapping cases but there might be other examples). I still need to dig around some more but I'm pretty sure all of these processes already only use bulk delete, and so won't have a chance to create them anyway. My instinct then is to not create a CommCareCaseManager.delete / bulk delete override that makes tombstoning the default and then exclude these processes. I think it might be better to be explicit about when bulk deletion should create tombstones to keep the DeletedSQLDoc table less cluttered with docs we'll probably never reference (unless it we potentially will?).

Reply:
> Case.delete() can't negatively impact tests because there's an explicit check that excludes calls from tests from making tombstones.

Yeah I understand that the check to exclude creating tombstones in tests exists, but I'm implying that we could potentially remove that check since these tombstones (DeletedSQLDoc objects) shouldn't actually cause issues in tests the way Domain-Deleted objects might have, which is especially true since SQL updates are automatically rolled back at the end of tests. Is there a scenario where it makes sense to actually use the `case.delete(leave_tombstone=False)` behavior in tests? Otherwise I think we can remove it and simplify this method.

As for the system forms/cases, it is true that I don't think we should create tombstones for the user owner mapping cases, but I'm not sure I'd extend that logic to other system forms. My reasoning is that the user owner mapping cases have a very specific workflow where cases are frequently hard deleted, and then re-created with the same case id (intentionally). As far as I know, there aren't any other forms/cases that behave this way. And fwiw, you can view the logic for deleting these cases which starts [here](https://github.com/dimagi/commcare-hq/blob/014db6c596a02fd02096a09b744a123212d11156/corehq/apps/users/models.py#L2201-L2208) and [ends up](https://github.com/dimagi/commcare-hq/blob/014db6c596a02fd02096a09b744a123212d11156/corehq/form_processor/models/cases.py#L280-L295) using the [hard_delete_cases](https://github.com/dimagi/commcare-hq/blob/014db6c596a02fd02096a09b744a123212d11156/corehq/sql_accessors/sql_templates/hard_delete_cases_2.sql) SQL function which I assume bypasses the tombstoning logic regardless of if it is defined only on CommCareCase.delete or on CommCareCaseManager.delete as well. 

Could be easiest to offline more about this in slack as well just to make sure we are all on the same page. Personally, I don't feel confident about what the right approach is here.

Reply:
> Is there a scenario where it makes sense to actually use the case.delete(leave_tombstone=False) behavior in tests? Otherwise I think we can remove it and simplify this method.

Not specifically (except the test I wrote to test that scenario) but I don't think creating tombstones on delete would be relevant to a lot of these tests either and at worst, I feel like all these sql updates could add up to really slow down testing as a whole.

> SQL function which I assume bypasses the tombstoning logic regardless of if it is defined only on CommCareCase.delete or on CommCareCaseManager.delete as well.

thanks for pointing this out for me!! you're right about the this bypassing tombstone creation for cases but it does look like it also eventually calls [hard_delete_cases_and_forms](https://github.com/dimagi/commcare-hq/blob/014db6c596a02fd02096a09b744a123212d11156/corehq/form_processor/backends/sql/processor.py#L63) > `hard_delete_forms` which does make use of the XFormInstanceManager.delete function so we'll have to exclude it explicitly there.

and sorry for the late response but after some more thinking I'm walking back on my last statement about not making tombstoning a default for bulk deletion, since for future work that makes use of it, it's probably better to err on the side of creating tombstones we don't really need than not having them. I'd be very open to offline about this later or do a short call though just to confirm what the next step should be.

Reply:
I think this thread is resolved now. @gherceg do you agree?

Reply:
Agreed